File: packages/skillet-eval-python/src/skill_eval/config/loader.py
from __future__ import annotations

import os
import sys
from pathlib import Path

import yaml

from .env import resolve_api_key
from .registry import DEFAULT_MODELS, ENV_KEY_MAP, infer_provider
from .schema import (
    CliOverrides,
    Config,
    GraderConfig,
    ProviderConfig,
    ResolvedConfig,
    ResolvedGraderConfig,
    ResolvedProviderConfig,
    SettingsConfig,
)


def _detect_providers() -> list[ProviderConfig]:
    providers: list[ProviderConfig] = []
    for name, env_names in ENV_KEY_MAP.items():
        for env_name in env_names:
            if os.environ.get(env_name):
                providers.append(
                    ProviderConfig(
                        name=name,
                        model=DEFAULT_MODELS[name],
                        api_key=os.environ[env_name],
                    )
                )
                break
    return providers


def load_config(overrides: CliOverrides | None = None) -> ResolvedConfig:
    overrides = overrides or CliOverrides()
    config_path = overrides.config_path or "skillet.eval.yaml"
    file_config: Config | None = None

    if Path(config_path).exists():
        raw = Path(config_path).read_text()
        parsed = yaml.safe_load(raw)
        file_config = Config.model_validate(parsed)

    providers = file_config.providers if file_config else _detect_providers()

    if overrides.providers:
        providers = [p for p in providers if p.name in overrides.providers]

    if overrides.models:
        resolved_providers: list[ProviderConfig] = []
        for spec in overrides.models:
            if ":" in spec:
                name, model = spec.split(":", 1)
            else:
                name = infer_provider(spec)
                model = spec
            key = resolve_api_key(name)
            resolved_providers.append(
                ProviderConfig(name=name, model=model, api_key=key)
            )
        providers = resolved_providers

    skipped: list[str] = []
    resolved: list[ResolvedProviderConfig] = []
    for p in providers:
        api_key = resolve_api_key(p.name, p.api_key)
        if not api_key:
            env_hint = ENV_KEY_MAP.get(p.name, [f"{p.name} API key"])[0]
            skipped.append(f"{p.model} (no {env_hint})")
            continue
        resolved.append(
            ResolvedProviderConfig(name=p.name, model=p.model, api_key=api_key)
        )

    if skipped:
        print(f"  ⚠ Skipped: {', '.join(skipped)}\n", file=sys.stderr)

    if not resolved:
        raise RuntimeError(
            "No providers configured. Set API keys in the environment "
            "(ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY) "
            "or create a skillet.eval.yaml config file."
        )

    grader_source = (
        file_config.grader
        if file_config and file_config.grader
        else GraderConfig(provider=resolved[0].name, model=resolved[0].model)
    )
    grader_key = resolve_api_key(grader_source.provider)
    if not grader_key:
        raise RuntimeError(
            f'No API key found for grader provider "{grader_source.provider}"'
        )

    settings = (
        file_config.settings
        if file_config
        else SettingsConfig()
    )
    if overrides.runs is not None:
        settings = SettingsConfig(
            max_steps=settings.max_steps,
            timeout=settings.timeout,
            runs_per_provider=overrides.runs,
            temperature=settings.temperature,
        )
    if overrides.timeout is not None:
        settings = SettingsConfig(
            max_steps=settings.max_steps,
            timeout=overrides.timeout,
            runs_per_provider=settings.runs_per_provider,
            temperature=settings.temperature,
        )

    grader = ResolvedGraderConfig(
        provider=grader_source.provider,
        model=grader_source.model,
        api_key=grader_key,
    )

    return ResolvedConfig(
        providers=resolved,
        grader=grader,
        settings=settings,
    )

File: packages/skillet/src/eval/config/loader.ts
import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";
import { resolveApiKey } from "./env.js";
import { DEFAULT_MODELS, ENV_KEY_MAP, inferProvider } from "./registry.js";
import {
	type CliOverrides,
	type Config,
	ConfigFileSchema,
	type ProviderConfig,
	type ResolvedConfig,
} from "./schema.js";

/** Auto-detect providers from environment variables. */
function detectProviders(): ProviderConfig[] {
	const providers: ProviderConfig[] = [];
	for (const [name, envNames] of Object.entries(ENV_KEY_MAP)) {
		for (const envName of envNames) {
			if (process.env[envName]) {
				providers.push({
					name: name as ProviderConfig["name"],
					model: DEFAULT_MODELS[name],
					apiKey: process.env[envName],
				});
				break;
			}
		}
	}
	return providers;
}

/** Coerce a short-form provider entry (model id string) into a full ProviderConfig. */
function normalizeProviderEntry(entry: string | ProviderConfig): ProviderConfig {
	if (typeof entry !== "string") return entry;
	const name = inferProvider(entry);
	return { name, model: entry };
}

/**
 * Load config from YAML file (if present), merge with env var auto-detection,
 * and apply CLI overrides. Returns a fully resolved config with API keys.
 */
export function loadConfig(overrides: CliOverrides = {}): ResolvedConfig {
	const configPath = overrides.configPath ?? "skillet.config.yaml";
	const resolvedConfigPath = path.resolve(configPath);
	let fileConfig: Config | undefined;

	if (overrides.configPath && !fs.existsSync(configPath)) {
		throw new Error(
			`Config file not found at ${configPath}.\nCheck the --config path, or run "skillet eval init" to scaffold one.`,
		);
	}

	if (fs.existsSync(configPath)) {
		const raw = fs.readFileSync(configPath, "utf-8");
		const parsed = YAML.parse(raw);
		fileConfig = ConfigFileSchema.parse(parsed);
	}

	const configDir = fileConfig ? path.dirname(resolvedConfigPath) : process.cwd();

	let providers: ProviderConfig[] =
		fileConfig?.providers.map(normalizeProviderEntry) ?? detectProviders();

	if (overrides.providers?.length) {
		providers = providers.filter((p) => overrides.providers?.includes(p.name));
	}

	if (overrides.models?.length) {
		type ProviderName = "anthropic" | "openai" | "google";
		providers = overrides.models.map((spec) => {
			let name: ProviderName;
			let model: string;

			if (spec.includes(":")) {
				const [n, m] = spec.split(":");
				name = n as ProviderName;
				model = m;
			} else {
				name = inferProvider(spec);
				model = spec;
			}

			return { name, model, apiKey: resolveApiKey(name) };
		});
	}

	const skipped: string[] = [];
	const resolved = providers
		.map((p) => {
			const apiKey = resolveApiKey(p.name, p.apiKey);
			if (!apiKey) {
				skipped.push(`${p.model} (no ${ENV_KEY_MAP[p.name]?.[0] ?? `${p.name} API key`})`);
				return null;
			}
			return { ...p, apiKey };
		})
		.filter((p): p is ProviderConfig & { apiKey: string } => p !== null);

	if (skipped.length > 0) {
		console.error(`  ⚠ Skipped: ${skipped.join(", ")}\n`);
	}

	if (resolved.length === 0) {
		throw new Error(
			"No providers configured. Set API keys in the environment " +
				"(ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY) " +
				"or create a skillet.config.yaml config file.",
		);
	}

	const graderSource = fileConfig?.grader ?? {
		provider: resolved[0].name,
		model: resolved[0].model,
	};
	const graderKey = resolveApiKey(graderSource.provider, undefined);
	if (!graderKey) {
		throw new Error(`No API key found for grader provider "${graderSource.provider}"`);
	}

	const settings = fileConfig?.settings ?? ConfigFileSchema.shape.settings.parse(undefined);
	if (overrides.runs) settings.runsPerProvider = overrides.runs;
	if (overrides.timeout) settings.timeout = overrides.timeout;

	const mocks = Object.fromEntries(
		Object.entries(fileConfig?.mocks ?? {}).map(([name, mock]) => {
			const resolveOne = (value: string) =>
				/^https?:\/\//i.test(value) ? value : path.resolve(configDir, value);
			const resolveMaybeMany = (value: string | string[] | undefined) => {
				if (!value) return undefined;
				return Array.isArray(value) ? value.map(resolveOne) : resolveOne(value);
			};
			return [
				name,
				{
					...mock,
					openapi: resolveMaybeMany(mock.openapi),
					mcpServer: resolveMaybeMany(mock.mcpServer),
				},
			];
		}),
	);

	return {
		providers: resolved,
		grader: { ...graderSource, apiKey: graderKey },
		skillRoots: (fileConfig?.skills.roots ?? []).map((root) => path.resolve(configDir, root)),
		settings,
		mocks,
	};
}
